Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut

```diff
@@ -1,86 +1,550 @@
+}
+
+void function NSSetModeCategory( string mode, int category )
+{
+	if( mode in file.categoryOverrides )
+	{
+		file.categoryOverrides[mode] = category
+		printt( "Overwriting category for mode:", mode )
+		return
+	}
+
+	file.categoryOverrides[mode] <- category
+}
+
+void function OnBtnFiltersClear_Activate( var b )
+{
+	file.searchString = ""
+	file.searchEnum = -1
 
-	AddMenuEventHandler( menu, eUIEvent.MENU_OPEN, OnOpenModesMenu )
+	SetConVarInt( "modemenu_mode_filter", -1 )
+	Hud_SetText( Hud_GetChild( file.menu, "BtnModeSearch"), "" )
 
-	AddEventHandlerToButtonClass( menu, "ModeButton", UIE_GET_FOCUS, ModeButton_GetFocus )
-	AddEventHandlerToButtonClass( menu, "ModeButton", UIE_CLICK, ModeButton_Click )
+	file.scrollOffset = 0
 
-	AddMenuFooterOption( menu, BUTTON_A, "#A_BUTTON_SELECT" )
-	AddMenuFooterOption( menu, BUTTON_B, "#B_BUTTON_BACK", "#BACK" )
-	
-	AddMenuFooterOption( menu, BUTTON_SHOULDER_LEFT, "#PRIVATE_MATCH_PAGE_PREV", "#PRIVATE_MATCH_PAGE_PREV", CycleModesBack, IsNorthstarServer )
-	AddMenuFooterOption( menu, BUTTON_SHOULDER_RIGHT, "#PRIVATE_MATCH_PAGE_NEXT", "#PRIVATE_MATCH_PAGE_NEXT", CycleModesForward, IsNorthstarServer )
+	BuildSortedModesArray()
+	UpdateListSliderHeight(float(file.sortedModes.len()))
+	UpdateListSliderPosition(file.sortedModes.len())
+	UpdateVisibleModes()
+}
+
+void function FilterAndUpdateList( var n )
+{
+	file.searchString = Hud_GetUTF8Text( Hud_GetChild( file.menu, "BtnModeSearch" ) )
+	file.searchEnum = GetConVarInt( "modemenu_mode_filter" )
+
+	file.scrollOffset = 0
+
+	BuildSortedModesArray()
+	UpdateListSliderHeight(float(file.sortedModes.len()))
+	UpdateListSliderPosition(file.sortedModes.len())
+	UpdateVisibleModes()
 }
 
 void function OnOpenModesMenu()
 {
+	RegisterButtonPressedCallback( MOUSE_WHEEL_UP , OnScrollUp )
+	RegisterButtonPressedCallback( MOUSE_WHEEL_DOWN , OnScrollDown )
+
+	// Reset filters
+	file.searchString = ""
+	file.searchEnum = -1
+
+	// We rebuild the modes array on open menu to make sure
+	// all modes get listed
+	BuildModesArray()
+	BuildSortedModesArray()
+
+	UpdateListSliderHeight(float(file.sortedModes.len()))
+	UpdateListSliderPosition(file.sortedModes.len())
 	UpdateVisibleModes()
-	
-	if ( level.ui.privatematch_mode == 0 ) // set to the first mode if there's no mode focused
-		Hud_SetFocused( GetElementsByClassname( GetMenu( "ModesMenu" ), "ModeButton" )[ 0 ] )
+
+	// Set to the first mode if there's no mode focused
+	if ( level.ui.privatematch_mode == 0 )
+	{
+		array<var> panels = GetElementsByClassname( file.menu, "ModeSelectorPanel" )
+		foreach( var panel in panels )
+		{
+			if( Hud_IsEnabled( Hud_GetChild( panel, "BtnMode") ) )
+			{
+				Hud_SetFocused( Hud_GetChild( panel, "BtnMode") )
+				break
+			}
+		}
+	}
+}
+
+void function OnCloseModesMenu()
+{
+	try
+	{
+		DeregisterButtonPressedCallback( MOUSE_WHEEL_UP , OnScrollUp )
+		DeregisterButtonPressedCallback( MOUSE_WHEEL_DOWN , OnScrollDown )
+	}
+	catch ( ex ) {}
```

Comment:
not a great solution

Reply:
tbh i dont even know why it needs to be in a try catch. Serverbrowser originally didnt have it and noone crashed?

Reply:
This crashes if the buttons aren't registered, but if your code isn't bad that should never happen